{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "# Run whole-cell simulation and parameter sweep reporter simulation\n",
    "\n",
    "<div id=\"toc-wrapper\"><h3> Table of Contents </h3><div id=\"toc\" style=\"max-height: 787px;\"><ol class=\"toc-item\"><li><a href=\"#Globals\">Globals</a></li><li><a href=\"#Get-codon-sequence-of-all-E.-coli-genes\">Get codon sequence of all  genes</a></li><li><a href=\"#Calculate-mRNA-copy-numbers-and-translation-initiation-rate-for-all-E.-coli-mRNAs\">Calculate mRNA copy numbers and translation initiation rate for all  mRNAs</a></li><li><a href=\"#Prepare-tRNA-input-files\">Prepare tRNA input files</a></li><li><a href=\"#Create-mRNA-sequence-file-for-leucine-starvation-whole-cell-simulation-(Run-1)\">Create mRNA sequence file for leucine starvation whole-cell simulation (Run 1)</a></li><li><a href=\"#Simulation-run-1:-Single-whole-cell-simulation-to-calculate-steady-state-charged-tRNA-fraction-during-Leu-starvation\">Simulation run 1: Single whole-cell simulation to calculate steady state charged tRNA fraction during Leu starvation</a></li><li><a href=\"#Create-mRNA-sequence-for-serine-starvation-whole-cell-simulation-(Run-12)\">Create mRNA sequence for serine starvation whole cell simulation (Run 12)</a></li><li><a href=\"#Simulation-run-12:-Single-whole-cell-simulation-to-calculate-steady-state-charged-tRNA-fraction-during-Ser-starvation\">Simulation run 12: Single whole-cell simulation to calculate steady state charged tRNA fraction during Ser starvation</a></li><li><a href=\"#Create-average-tRNA-concentration-file-during-leucine-starvation-based-on-whole-cell-simulation-Run-1\">Create average tRNA concentration file during leucine starvation based on whole-cell simulation Run 1</a></li><li><a href=\"#Create-average-tRNA-concentration-file-during-serine-starvation-based-on-whole-cell-simulation-Run-12\">Create average tRNA concentration file during serine starvation based on whole-cell simulation Run 12</a></li><li><a href=\"#Create-mRNA-sequences-for-simulation-run-2-with-systematically-varying-tRNA-accommodation-rate-(stall-duration)-at-Leu-codons\">Create mRNA sequences for simulation run 2 with systematically varying tRNA accommodation rate (stall duration) at Leu codons</a></li><li><a href=\"#Create-varying-tRNA-accommodation-rate-(stall-duration)-at-Leu-codons-for-simulation-run-2\">Create varying tRNA accommodation rate (stall duration) at Leu codons for simulation run 2</a></li><li><a href=\"#Simulation-run-2:-Reporter-simulation-with-systematically-varying-duration-of-ribosome-stalling-during-leucine-starvation\">Simulation run 2: Reporter simulation with systematically varying duration of ribosome stalling during leucine starvation</a></li><li><a href=\"#Create-mRNA-sequences-for-simulation-run-13-with-systematically-varying-tRNA-accommodation-rate-(stall-duration)-at-Ser-codons\">Create mRNA sequences for simulation run 13 with systematically varying tRNA accommodation rate (stall duration) at Ser codons</a></li><li><a href=\"#Create-varying-tRNA-accommodation-rate-(stall-duration)-at-Ser-codons-for-simulation-run-13\">Create varying tRNA accommodation rate (stall duration) at Ser codons for simulation run 13</a></li><li><a href=\"#Simulation-run-13:-Reporter-simulation-with-systematically-varying-duration-of-ribosome-stalling-during-serine-starvation\">Simulation run 13: Reporter simulation with systematically varying duration of ribosome stalling during serine starvation</a></li></ol></div></div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "### Globals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "# sequence input and output\n",
    "import Bio.SeqIO\n",
    "# provides dictionary of codon names\n",
    "from Bio.SeqUtils.CodonUsage import SynonymousCodons\n",
    "# for converting 3 letter amino acid code to 1 letter code\n",
    "from Bio.SeqUtils import seq1\n",
    "# for fast access of fasta files\n",
    "import pyfaidx\n",
    "# for parsing GFF3 files\n",
    "import HTSeq\n",
    "# for tab data processing\n",
    "import pandas as pd\n",
    "# numeric and matrix library\n",
    "import numpy as np\n",
    "# shell utilities\n",
    "import shutil\n",
    "# for submitting shell commands\n",
    "import subprocess as sp\n",
    "# for string matching\n",
    "import re\n",
    "\n",
    "# create a dictionary of codon names and number (arranged alphabetically by aa)\n",
    "codonnum = 0\n",
    "codonDict = dict()\n",
    "for aa in sorted(SynonymousCodons, key=lambda aa3: seq1(aa3)):\n",
    "    if aa == 'STOP':\n",
    "        continue\n",
    "    for codon in sorted(SynonymousCodons[aa]):\n",
    "        # these two codons are numbered out of order for consistent notation\n",
    "        # with Subramaniam et al. Cell 2014\n",
    "        if codon in ['AGC']:\n",
    "            codonDict['AGC'] = 59\n",
    "        elif codon in ['AGT']:\n",
    "            codonDict['AGT'] = 60\n",
    "        else:\n",
    "            codonDict[codon] = codonnum\n",
    "            codonnum += 1\n",
    "\n",
    "# to convert 3 letter codons to numbers between 0 and 63\n",
    "\n",
    "\n",
    "def get_numerical_codon_sequence(seq):\n",
    "    numseq = list()\n",
    "    for pos in range(0, len(seq) - 3, 3):\n",
    "        try:\n",
    "            numseq.append(str(codonDict[seq[pos:pos + 3]]))\n",
    "        except KeyError:\n",
    "            numseq.append('-1')\n",
    "            raise\n",
    "            return None\n",
    "    return ' '.join(numseq)\n",
    "\n",
    "\n",
    "# starting yfp sequence for leucine starvation expts\n",
    "yfp0 = Bio.SeqIO.read('../annotations/simulations/yfp0.fa', 'fasta')\n",
    "yfp0 = str(yfp0.seq)\n",
    "\n",
    "# starting sequence for serine starvation expts\n",
    "# all ser codons in yfp0 were AGC\n",
    "yfp_agc = list(yfp0)\n",
    "for pos in range(0, len(yfp0), 3):\n",
    "    current_codon = yfp0[pos:pos + 3]\n",
    "    if current_codon in SynonymousCodons['SER']:\n",
    "        yfp_agc[pos:pos + 3] = 'AGC'\n",
    "yfp_agc = ''.join(yfp_agc)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "### Get codon sequence of all *E. coli* genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "genome = '../annotations/simulations/NC_000913.fna'\n",
    "annotations = '../annotations/simulations/NC_000913.gff'\n",
    "\n",
    "# read genome as a python dictionary\n",
    "genome = pyfaidx.Fasta(genome)\n",
    "\n",
    "annotationrecords = HTSeq.GFF_Reader(annotations)\n",
    "\n",
    "annotationdf = dict()\n",
    "for record in annotationrecords:\n",
    "    if record.type != 'CDS':\n",
    "        continue\n",
    "    if 'pseudo' in record.attr.keys() and record.attr['pseudo'] == 'true':\n",
    "        continue\n",
    "    # fdnG, fdoG, fdhF have TGA encoded selenocysteine codons\n",
    "    if record.attr['gene'] in ['fdnG', 'fdoG', 'fdhF']:\n",
    "        continue\n",
    "    sequence = str(genome[record.iv.chrom][record.iv.start:record.iv.end])\n",
    "    if record.iv.strand == '-':\n",
    "        sequence = str(HTSeq.Sequence(sequence).get_reverse_complement())\n",
    "\n",
    "    annotationdf[record.attr['ID']] = {\n",
    "        'iv': record.iv,\n",
    "        'gene': record.attr['gene'],\n",
    "        'product': record.attr['product'],\n",
    "        'start': record.iv.start,\n",
    "        'end': record.iv.end,\n",
    "        'strand': record.iv.strand,\n",
    "        'sequence': sequence,\n",
    "    }\n",
    "annotationdf = pd.DataFrame.from_dict(annotationdf, orient='index')\n",
    "annotationdf['length'] = annotationdf['iv'].apply(lambda x: x.length)\n",
    "\n",
    "annotationdf['numsequence'] = annotationdf['sequence'].apply(\n",
    "    get_numerical_codon_sequence)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "### Calculate mRNA copy numbers and translation initiation rate for all *E. coli* mRNAs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "# Number of ribosomes (molecules per cell), 37C, 20 min doubling\n",
    "# Bremer 2008\n",
    "nRibo = 73000\n",
    "\n",
    "# Number of mrna (nt per cell), 37C, 20 min doubling\n",
    "# Bremer 2008\n",
    "nMrna = 4.3e6\n",
    "\n",
    "# Bound fraction of ribosomes, 37C, 20 min doubling\n",
    "# Bremer 2008\n",
    "boundRiboFract = 0.85\n",
    "\n",
    "# Mean elongation rate (s^-1), 37C, 20 min doubling\n",
    "# Bremer 2008\n",
    "meanElongationRate = 22\n",
    "\n",
    "rawdata = pd.read_table(\n",
    "    '../annotations/simulations/ecoli.mrna.concn.and.te.li.2014.csv',\n",
    "    sep=',', )\n",
    "\n",
    "rawdata = rawdata.dropna().set_index('Gene')\n",
    "rawdata['mRNA level (RPKM)'] = rawdata['mRNA level (RPKM)'].apply(int)\n",
    "combined = annotationdf.join(rawdata, on='gene', how='inner')\n",
    "\n",
    "# the total number of mrnaribonucleotides in the cell should be equal to\n",
    "# the number of each mrna species multiplied by its length\n",
    "\n",
    "combined['rpkm_nt'] = combined['mRNA level (RPKM)'] * combined['length']\n",
    "\n",
    "mrnaNormalization = combined['rpkm_nt'].sum() / nMrna\n",
    "\n",
    "combined['mrnaCopyNumber'] = combined['mRNA level (RPKM)'].apply(\n",
    "    lambda x: int(np.round(x / mrnaNormalization)))\n",
    "\n",
    "combined = combined[combined.mrnaCopyNumber > 0]\n",
    "\n",
    "# basic idea for TE normalization below is the equality:\n",
    "# total number of mrna bound ribosomes in the cell =\n",
    "# sum over all mrnas( initation rate * mrna copy number\n",
    "#                                       * length / elongation rate)\n",
    "\n",
    "combined['boundribosomes'] = (combined['length'] *\n",
    "                              combined['Translation efficiency (AU)'] *\n",
    "                              combined['mrnaCopyNumber'])\n",
    "\n",
    "initiationRateNormalization = (combined['boundribosomes'].sum() /\n",
    "                               (nRibo * boundRiboFract * meanElongationRate))\n",
    "\n",
    "combined['initiationRate'] = (\n",
    "    combined['Translation efficiency (AU)'] /\n",
    "    initiationRateNormalization).apply(lambda x: np.round(x, 4))\n",
    "\n",
    "# Write initiation rate, mRNA copy number and the codon sequence to a file\n",
    "# for using in the simulation\n",
    "\n",
    "temp = combined[['initiationRate', 'mrnaCopyNumber', 'numsequence'\n",
    "                 ]].sort_values(\n",
    "                     by=['mrnaCopyNumber'], ascending=False)\n",
    "temp.to_csv(\n",
    "    '../annotations/simulations/run1/run1_ecoli_mrnas.csv',\n",
    "    sep='\\t',\n",
    "    index=False,\n",
    "    header=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "### Prepare tRNA input files\n",
    "\n",
    "0. tRNA concentration\n",
    "0. tRNA-codon cognate pairs\n",
    "0. tRNA-codon cognate weights\n",
    "0. tRNA aminoacylation rate constants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "max_aminoacylation_rate = 2.0e10\n",
    "\n",
    "# Read tRNA concentrations, anticodon, cognate codons\n",
    "# from Table 2 in Dong et al. J. Mol. Biol. 1996.\n",
    "# Gly1 and Ile2 entries were manually deleted from the table\n",
    "# since these were not measured in this study.\n",
    "# Sec tRNA was deleted since we are not considering\n",
    "# this non-canonical translation.\n",
    "# A typo for Val1 anticodon was corrected (TAG â†’ TAC).\n",
    "lines = open('../annotations/simulations/ecoli.trna.abundance.dong1996.txt'\n",
    "             ).read().splitlines()\n",
    "trnas = dict()\n",
    "# read every 6th line until last but one.\n",
    "for trnaindex, trna in enumerate(lines[0:-1:6]):\n",
    "    trnas[trna] = dict()\n",
    "    trnas[trna]['anticodon'] = lines[trnaindex * 6 + 1].strip().replace('U',\n",
    "                                                                        'T')\n",
    "    trnas[trna]['codons'] = [\n",
    "        codon.strip().replace('U', 'T')\n",
    "        for codon in lines[trnaindex * 6 + 2].split(',')\n",
    "    ]\n",
    "    trnas[trna]['mean_concn'] = int(lines[trnaindex * 6 + 4].split('(')[0])\n",
    "    trnas[trna]['std_concn'] = int(\n",
    "        lines[trnaindex * 6 + 4].split('(')[1].split(')')[0].strip())\n",
    "    trnas[trna]['fraction'] = float(lines[trnaindex * 6 + 5].strip())\n",
    "    trnas[trna]['aminoacid'] = trna[:3].lower()\n",
    "\n",
    "# Ile2 was not measured in Dong 1996. So assigning it to Ile1.\n",
    "trnas['Ile1']['codons'].append('ATA')\n",
    "\n",
    "# Calculate the factors for codon-tRNA interaction\n",
    "# that multiply the kcat/KM for codon-tRNA reading:\n",
    "# w = 1 for Watson-Crick base pairs,\n",
    "# w = 0.64 for purine-pyrimidine mismatch,\n",
    "# w = 0.61 for purine-purine mismatch. (Weights based on Shah 2013).\n",
    "\n",
    "# If two trnas have the same anticodon\n",
    "# they are considered as a single trna species in the simulation.\n",
    "concentrations = dict()\n",
    "cognate_pairings = dict()\n",
    "weights = dict()\n",
    "\n",
    "for trna in trnas:\n",
    "    cognatecodons = trnas[trna]['codons']\n",
    "    if trnas[trna]['anticodon'] not in concentrations:\n",
    "        concentrations[trnas[trna]['anticodon']] = trnas[trna]['mean_concn']\n",
    "    else:\n",
    "        concentrations[trnas[trna]['anticodon']] += trnas[trna]['mean_concn']\n",
    "\n",
    "    for codon in cognatecodons:\n",
    "        # pairing at the wobble position.\n",
    "        pairing = codon[-1:] + trnas[trna]['anticodon'][:1]\n",
    "\n",
    "        watson_crick_pairs = ['GC', 'AT', 'TA', 'CG']\n",
    "        pur_pyr_pairs = ['GT', 'TG', 'CA']\n",
    "        pur_pur_pairs = ['TT', 'AA']\n",
    "\n",
    "        if pairing in watson_crick_pairs:\n",
    "            weight = 1\n",
    "        elif pairing in pur_pyr_pairs:\n",
    "            weight = 0.64\n",
    "        else:\n",
    "            weight = 0.61\n",
    "\n",
    "        if codon not in cognate_pairings:\n",
    "            cognate_pairings[codon] = [trnas[trna]['anticodon']]\n",
    "            weights[codon] = [weight]\n",
    "        elif trnas[trna]['anticodon'] not in cognate_pairings[codon]:\n",
    "            cognate_pairings[codon].append(trnas[trna]['anticodon'])\n",
    "            weights[codon].append(weight)\n",
    "\n",
    "# trnaindices are sorted alphabetically.\n",
    "\n",
    "# concentrations file.\n",
    "outputFile = open(\n",
    "    \"../annotations/simulations/wholecell.trna.concentrations.tsv\", \"write\")\n",
    "outputFile.write('trnaindex\\tanticodon\\tmolpercell\\taminoacid\\n')\n",
    "# trnaindex 0 is a dummy trna without any anticodon.\n",
    "outputFile.write('0\\tNNN\\t0\\txxx\\n')\n",
    "trnaindices = dict()\n",
    "for trnaindex, anticodon in enumerate(sorted(concentrations)):\n",
    "    aminoacid = [\n",
    "        trnas[trna]['aminoacid'] for trna in trnas\n",
    "        if trnas[trna]['anticodon'] == anticodon\n",
    "    ][0]\n",
    "    trnaindices[anticodon] = trnaindex + 1\n",
    "    outputFile.write(\n",
    "        str(trnaindex + 1) + '\\t' + anticodon + '\\t' + '%d' %\n",
    "        concentrations[anticodon] + '\\t' + aminoacid + '\\n')\n",
    "outputFile.close()\n",
    "\n",
    "# aminoacylation rates file.\n",
    "outputFile = open(\n",
    "    \"../annotations/simulations/wholecell.trna.aminoacylation.rate.tsv\",\n",
    "    \"write\")\n",
    "outputFile.write('trnaindex\\tanticodon\\tkcatBykm\\n')\n",
    "outputFile.write('0\\tNNN\\t0\\n')\n",
    "trnaindices = dict()\n",
    "for trnaindex, anticodon in enumerate(sorted(concentrations)):\n",
    "    trnaindices[anticodon] = trnaindex + 1\n",
    "    outputFile.write(\n",
    "        str(trnaindex + 1) + '\\t' + anticodon + '\\t' + '%d' %\n",
    "        max_aminoacylation_rate + '\\n')\n",
    "outputFile.close()\n",
    "\n",
    "# cognate pairings file.\n",
    "outputFile = open(\"../annotations/simulations/wholecell.cognate.pairs.tsv\",\n",
    "                  \"write\")\n",
    "outputFile.write('codonindex\\tcodon\\ttrnaindex1\\ttrnaindex2\\n')\n",
    "for codon in sorted(codonDict, key=lambda x: codonDict[x]):\n",
    "    if len(cognate_pairings[codon]) == 1:\n",
    "        outputFile.write(\n",
    "            str(codonDict[codon]) + '\\t' + codon + '\\t' + '%d' %\n",
    "            trnaindices[cognate_pairings[codon][0]] + '\\t0\\n')\n",
    "    elif len(cognate_pairings[codon]) == 2:\n",
    "        outputFile.write(\n",
    "            str(codonDict[codon]) + '\\t' + codon + '\\t' + '%d' %\n",
    "            trnaindices[cognate_pairings[codon][0]] + '\\t' + '%d' %\n",
    "            trnaindices[cognate_pairings[codon][1]] + '\\n')\n",
    "outputFile.close()\n",
    "\n",
    "# cognate weights file.\n",
    "outputFile = open(\"../annotations/simulations/wholecell.cognate.weights.tsv\",\n",
    "                  \"write\")\n",
    "outputFile.write('codonindex\\tcodon\\ttrnaweight1\\ttrnaweight2\\n')\n",
    "for codon in sorted(codonDict, key=lambda x: codonDict[x]):\n",
    "    if len(weights[codon]) == 1:\n",
    "        outputFile.write(\n",
    "            str(codonDict[codon]) + '\\t' + codon + '\\t' + '%0.2f' %\n",
    "            weights[codon][0] + '\\t0.00\\n')\n",
    "    elif len(weights[codon]) == 2:\n",
    "        outputFile.write(\n",
    "            str(codonDict[codon]) + '\\t' + codon + '\\t' + '%0.2f' %\n",
    "            weights[codon][0] + '\\t' + '%0.2f' % weights[codon][1] + '\\n')\n",
    "outputFile.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "### Create mRNA sequence file for leucine starvation whole-cell simulation (Run 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "mutation_locations = [\n",
    "    {\n",
    "        15: 'cta'\n",
    "    },\n",
    "    {\n",
    "        10: 'cta',\n",
    "        14: 'cta',\n",
    "        15: 'cta'\n",
    "    },\n",
    "]\n",
    "\n",
    "yfpmutants = dict()\n",
    "for mutant in mutation_locations:\n",
    "    key = '_'.join(['yfp'] + [\n",
    "        codon + str(location) for location, codon in mutant.items()\n",
    "    ])\n",
    "    yfpmutants[key] = list(yfp0)\n",
    "    leucodon_number = 0\n",
    "    for position in range(0, len(yfp0), 3):\n",
    "        currentcodon = yfp0[position:position + 3]\n",
    "        # proceed only if the codon is a Leu codon (which are all CTG in yfp0)\n",
    "        if currentcodon not in ['CTG']:\n",
    "            continue\n",
    "        leucodon_number += 1\n",
    "        for location in mutant.keys():\n",
    "            if leucodon_number == location:\n",
    "                yfpmutants[key][position:position + 3] = mutant[\n",
    "                    location].upper()\n",
    "    yfpmutants[key] = ''.join(yfpmutants[key])\n",
    "\n",
    "# mrna copy number is arbitrary, but kept low so that\n",
    "# ribosomes are not overloaded\n",
    "defaultMrnaCopyNumber = 10  # per cell\n",
    "# median initiation rate of all E. coli mRNAs in the simulation, s-1\n",
    "defaultInitationRate = 0.3  # s-1,\n",
    "\n",
    "listOfInitiationRates = [defaultInitationRate]\n",
    "\n",
    "outputFile = '../annotations/simulations/run1/run1_ecoli_and_reporter_mrnas.csv'\n",
    "shutil.copyfile('../annotations/simulations/run1/run1_ecoli_mrnas.csv',\n",
    "                outputFile)\n",
    "\n",
    "for initiationRate in listOfInitiationRates:\n",
    "    File = open(outputFile, 'a')\n",
    "    File.write(\"%0.3f\\t%d\\t%s\\n\" % (initiationRate, defaultMrnaCopyNumber,\n",
    "                                    get_numerical_codon_sequence(yfp0[:-3])))\n",
    "    for mutant in sorted(yfpmutants, reverse=True):\n",
    "        num_seq = ''.join(\n",
    "            get_numerical_codon_sequence(yfpmutants[mutant][:-3]))\n",
    "        File.write(\"%0.3f\\t%d\\t%s\\n\" % (initiationRate, defaultMrnaCopyNumber,\n",
    "                                        num_seq))\n",
    "    File.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Simulation run 1: Single whole-cell simulation to calculate steady state charged tRNA fraction during Leu starvation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting simulation_run_1.py\n"
     ]
    }
   ],
   "source": [
    "%%writefile simulation_run_1.py\n",
    "#!/usr/bin/env python\n",
    "# SBATCH --mem=8000\n",
    "\n",
    "import subprocess as sp\n",
    "import sys\n",
    "\n",
    "jobindex = int(sys.argv[1])\n",
    "currentindex = -1\n",
    "\n",
    "for starvationfactor in [0.01]:\n",
    "    currentindex += 1\n",
    "    if currentindex != jobindex:\n",
    "        continue\n",
    "    sp.check_output(' '.join([\n",
    "        './wholecell_simulation',\n",
    "        '--threshold-time', '100',\n",
    "        '--total-time', '150',\n",
    "        '--aminoacid', 'leu',\n",
    "        'starvationfactor', '%0.2f' % starvationfactor,\n",
    "        '--aminoacid', 'leu',\n",
    "        'relativetrnaaminoacylationrates', '1', '2.2', '1.2', '1', '0.1',\n",
    "        '--output-prefix', '../rawdata/simulations/run1/',\n",
    "        '--input-genes', '../annotations/simulations/run1/run1_ecoli_and_reporter_mrnas.csv'\n",
    "    ]), shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import subprocess as sp\n",
    "# loop submits each simulation to a different node of the cluster\n",
    "for index in range(1):\n",
    "    sp.check_output([\n",
    "        'sbatch',  # for SLURM cluster; this line can be commented out if running locally\n",
    "        '-t',\n",
    "        '15',  # for SLURM cluster; this line can be commented out if running locally\n",
    "        '-n',\n",
    "        '1',  # for SLURM cluster; this line can be commented out if running locally\n",
    "        'simulation_run_1.py',\n",
    "        str(index)\n",
    "    ])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create mRNA sequence for serine starvation whole cell simulation (Run 12)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "aa = 'SER'\n",
    "\n",
    "mutation_locations = [{\n",
    "    2: 'tcg',\n",
    "    3: 'tcg',\n",
    "    4: 'tcg',\n",
    "    5: 'tcg',\n",
    "    6: 'tcg',\n",
    "    7: 'tcg',\n",
    "    8: 'tcg'\n",
    "}, ]\n",
    "\n",
    "yfpmutants = dict()\n",
    "for mutant in mutation_locations:\n",
    "    key = '_'.join(['yfp'] + [\n",
    "        codon + str(location) for location, codon in mutant.items()\n",
    "    ])\n",
    "    yfpmutants[key] = list(yfp_agc)\n",
    "    codon_number = 0\n",
    "    for position in range(0, len(yfp_agc), 3):\n",
    "        currentcodon = yfp_agc[position:position + 3]\n",
    "        # proceed only if the codon is a Leu codon (which are all CTG in yfp0)\n",
    "        if currentcodon not in SynonymousCodons[aa]:\n",
    "            continue\n",
    "        codon_number += 1\n",
    "        for location in mutant.keys():\n",
    "            if codon_number == location:\n",
    "                yfpmutants[key][position:position + 3] = mutant[\n",
    "                    location].upper()\n",
    "    yfpmutants[key] = ''.join(yfpmutants[key])\n",
    "\n",
    "# mrna copy number is arbitrary, but kept low so that\n",
    "# ribosomes are not overloaded\n",
    "defaultMrnaCopyNumber = 10  # per cell\n",
    "# median initiation rate of all E. coli mRNAs in the simulation, s-1\n",
    "defaultInitationRate = 0.3  # s-1,\n",
    "\n",
    "listOfInitiationRates = [defaultInitationRate]\n",
    "\n",
    "outputFile = '../annotations/simulations/run12/run12_ecoli_and_reporter_mrnas.csv'\n",
    "shutil.copyfile('../annotations/simulations/run1/run1_ecoli_mrnas.csv',\n",
    "                outputFile)\n",
    "\n",
    "for initiationRate in listOfInitiationRates:\n",
    "    File = open(outputFile, 'a')\n",
    "    File.write(\"%0.3f\\t%d\\t%s\\n\" %\n",
    "               (initiationRate, defaultMrnaCopyNumber,\n",
    "                get_numerical_codon_sequence(yfp_agc[:-3])))\n",
    "    for mutant in sorted(yfpmutants, reverse=True):\n",
    "        num_seq = ''.join(\n",
    "            get_numerical_codon_sequence(yfpmutants[mutant][:-3]))\n",
    "        File.write(\"%0.3f\\t%d\\t%s\\n\" % (initiationRate, defaultMrnaCopyNumber,\n",
    "                                        num_seq))\n",
    "    File.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Simulation run 12: Single whole-cell simulation to calculate steady state charged tRNA fraction during Ser starvation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting simulation_run_12.py\n"
     ]
    }
   ],
   "source": [
    "%%writefile simulation_run_12.py\n",
    "#!/usr/bin/env python\n",
    "# SBATCH --mem=8000\n",
    "\n",
    "import subprocess as sp\n",
    "import sys\n",
    "\n",
    "jobindex = int(sys.argv[1])\n",
    "currentindex = -1\n",
    "\n",
    "for starvationfactor in [0.01]:\n",
    "    currentindex += 1\n",
    "    if currentindex != jobindex:\n",
    "        continue\n",
    "    sp.check_output(' '.join([\n",
    "        './wholecell_simulation',\n",
    "        '--threshold-time', '100',\n",
    "        '--total-time', '150',\n",
    "        '--aminoacid', 'ser',\n",
    "        'starvationfactor', '%0.2f' % starvationfactor,\n",
    "        '--aminoacid', 'ser',\n",
    "        'relativetrnaaminoacylationrates', '0.1', '1.5', '1.5', '1',\n",
    "        '--output-prefix', '../rawdata/simulations/run12/',\n",
    "        '--input-genes', '../annotations/simulations/run12/run12_ecoli_and_reporter_mrnas.csv'\n",
    "    ]), shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import subprocess as sp\n",
    "# loop submits each simulation to a different node of the cluster\n",
    "for index in range(1):\n",
    "    sp.check_output([\n",
    "        'sbatch',  # for SLURM cluster; this line can be commented out if running locally\n",
    "        '-t',\n",
    "        '15',  # for SLURM cluster; this line can be commented out if running locally\n",
    "        '-n',\n",
    "        '1',  # for SLURM cluster; this line can be commented out if running locally\n",
    "        'simulation_run_12.py',\n",
    "        str(index)\n",
    "    ])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "### Create average tRNA concentration file during leucine starvation based on whole-cell simulation Run 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "inputfile = '../rawdata/simulations/run1/mrnafile_run1_ecoli_and_reporter_mrnas_thresholdtime_100_totaltime_150_leu_starvationfactor_0.01_leu_relativetrnaaminoacylationrates_1_2.2_1.2_1_0.1/avg_ribo_tRNA.out'\n",
    "trnaconcentrationfile = pd.read_table(\n",
    "    '../annotations/simulations/wholecell.trna.concentrations.tsv')\n",
    "trnaconcnlist = open(inputfile).readlines()\n",
    "trnaConcn = list()\n",
    "for trnatype in ['acylated']:\n",
    "    temp = map(\n",
    "        lambda x: re.findall('^' + trnatype + '_trna_(\\d+)_([ACTG]{3})\\t([\\w\\.]+)', x),\n",
    "        trnaconcnlist)\n",
    "    temp = filter(lambda x: len(x), temp)\n",
    "    temp = pd.DataFrame(\n",
    "        [x[0] for x in temp], columns=['trnaindex', 'anticodon', 'molpercell'])\n",
    "    trnaConcn.append(temp)\n",
    "trnaConcn = pd.concat(trnaConcn, join='inner', axis=1)\n",
    "trnaConcn = trnaConcn.merge(\n",
    "    trnaconcentrationfile[['anticodon', 'aminoacid']], how='right').fillna(0)\n",
    "trnaConcn['trnaindex'] = trnaConcn['trnaindex'].apply(int)\n",
    "trnaConcn['molpercell'] = trnaConcn['molpercell'].apply(\n",
    "    lambda x: int(float(x)))\n",
    "trnaConcn.sort_values(by='trnaindex').to_csv(\n",
    "    '../annotations/simulations/leucine.starvation.average.trna.concentrations.tsv',\n",
    "    index=False,\n",
    "    sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create average tRNA concentration file during serine starvation based on whole-cell simulation Run 12"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "inputfile = (\n",
    "    '../rawdata/simulations/run12/mrnafile_run12_ecoli_and_reporter_mrnas_'\n",
    "    'thresholdtime_100_totaltime_150_ser_starvationfactor_0.01_'\n",
    "    'ser_relativetrnaaminoacylationrates_0.1_1.5_1.5_1/avg_ribo_tRNA.out')\n",
    "trnaconcentrationfile = pd.read_table(\n",
    "    '../annotations/simulations/wholecell.trna.concentrations.tsv')\n",
    "trnaconcnlist = open(inputfile).readlines()\n",
    "trnaConcn = list()\n",
    "for trnatype in ['acylated']:\n",
    "    temp = map(\n",
    "        lambda x: re.findall('^' + trnatype + '_trna_(\\d+)_([ACTG]{3})\\t([\\w\\.]+)', x),\n",
    "        trnaconcnlist)\n",
    "    temp = filter(lambda x: len(x), temp)\n",
    "    temp = pd.DataFrame(\n",
    "        [x[0] for x in temp], columns=['trnaindex', 'anticodon', 'molpercell'])\n",
    "    trnaConcn.append(temp)\n",
    "trnaConcn = pd.concat(trnaConcn, join='inner', axis=1)\n",
    "trnaConcn = trnaConcn.merge(\n",
    "    trnaconcentrationfile[['anticodon', 'aminoacid']], how='right').fillna(0)\n",
    "trnaConcn['trnaindex'] = trnaConcn['trnaindex'].apply(int)\n",
    "trnaConcn['molpercell'] = trnaConcn['molpercell'].apply(\n",
    "    lambda x: int(float(x)))\n",
    "trnaConcn.sort_values(by='trnaindex').to_csv(\n",
    "    '../annotations/simulations/serine.starvation.average.trna.concentrations.tsv',\n",
    "    index=False,\n",
    "    sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "### Create mRNA sequences for simulation run 2 with systematically varying tRNA accommodation rate (stall duration) at Leu codons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "mutation_locations = [\n",
    "    {\n",
    "        'cta': 2\n",
    "    },\n",
    "    {\n",
    "        'cta': 6\n",
    "    },\n",
    "    {\n",
    "        'cta': 8\n",
    "    },\n",
    "    {\n",
    "        'cta': 9\n",
    "    },\n",
    "    {\n",
    "        'cta': 10\n",
    "    },\n",
    "    {\n",
    "        'cta': 11\n",
    "    },\n",
    "    {\n",
    "        'cta': 12\n",
    "    },\n",
    "    {\n",
    "        'cta': 13\n",
    "    },\n",
    "    {\n",
    "        'cta': 14\n",
    "    },\n",
    "    {\n",
    "        'cta': 18\n",
    "    },\n",
    "    {\n",
    "        'ctc': 2\n",
    "    },\n",
    "    {\n",
    "        'ctc': 6\n",
    "    },\n",
    "    {\n",
    "        'ctc': 8\n",
    "    },\n",
    "    {\n",
    "        'ctc': 9\n",
    "    },\n",
    "    {\n",
    "        'ctc': 10\n",
    "    },\n",
    "    {\n",
    "        'ctc': 11\n",
    "    },\n",
    "    {\n",
    "        'ctc': 12\n",
    "    },\n",
    "    {\n",
    "        'ctc': 13\n",
    "    },\n",
    "    {\n",
    "        'ctc': 14\n",
    "    },\n",
    "    {\n",
    "        'ctc': 18\n",
    "    },\n",
    "    {\n",
    "        'ctt': 2\n",
    "    },\n",
    "    {\n",
    "        'ctt': 6\n",
    "    },\n",
    "    {\n",
    "        'ctt': 10\n",
    "    },\n",
    "    {\n",
    "        'ctt': 14\n",
    "    },\n",
    "    {\n",
    "        'ctt': 18\n",
    "    },\n",
    "]\n",
    "\n",
    "yfpmutants = dict()\n",
    "for mutant in mutation_locations:\n",
    "    key = '_'.join(['yfp'] + [\n",
    "        codon + str(location) for codon, location in mutant.items()\n",
    "    ])\n",
    "    yfpmutants[key] = list(yfp0)\n",
    "    leucodon_number = 0\n",
    "    for position in range(0, len(yfp0), 3):\n",
    "        currentcodon = yfp0[position:position + 3]\n",
    "        # proceed only if the codon is a Leu codon (which are all CTG in yfp0)\n",
    "        if currentcodon not in ['CTG']:\n",
    "            continue\n",
    "        leucodon_number += 1\n",
    "        for codon in mutant.keys():\n",
    "            if leucodon_number == mutant[codon]:\n",
    "                yfpmutants[key][position:position + 3] = codon.upper()\n",
    "    yfpmutants[key] = ''.join(yfpmutants[key])\n",
    "\n",
    "defaultMrnaCopyNumber = 1  # per cell\n",
    "defaultInitationRate = 0.3  # s-1, This is the median initiation rate\n",
    "\n",
    "listOfInitiationRates = [defaultInitationRate]\n",
    "\n",
    "for initiationRate in listOfInitiationRates:\n",
    "    for mutant in yfpmutants:\n",
    "        outputFile = '../annotations/simulations/run2/' + \\\n",
    "            '%s_initiationrate_%0.4f.csv' % (mutant, initiationRate)\n",
    "        num_seq = ''.join(\n",
    "            get_numerical_codon_sequence(yfpmutants[mutant][:-3]))\n",
    "        File = open(outputFile, 'w')\n",
    "\n",
    "        File.write(\"%0.4f\\t%d\\t%s\\n\" %\n",
    "                   (initiationRate, defaultMrnaCopyNumber,\n",
    "                    get_numerical_codon_sequence(yfp0[:-3])))\n",
    "\n",
    "        File.write(\"%0.4f\\t%d\\t%s\\n\" % (initiationRate, defaultMrnaCopyNumber,\n",
    "                                        num_seq))\n",
    "    File.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "### Create varying tRNA accommodation rate (stall duration) at Leu codons for simulation run 2 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "stallstrengthranges = {\n",
    "    'trafficjam': {\n",
    "        'cta': np.linspace(\n",
    "            0.01, 0.2, num=20),\n",
    "        'ctc': np.linspace(\n",
    "            0.02, 0.4, num=20),\n",
    "        'ctt': np.linspace(\n",
    "            0.02, 0.4, num=20),\n",
    "    },\n",
    "    '5primepreterm': {\n",
    "        'cta': np.linspace(\n",
    "            0.02, 0.4, num=20),\n",
    "        'ctc': np.linspace(\n",
    "            0.05, 1, num=20),\n",
    "        'ctt': np.linspace(\n",
    "            0.05, 1, num=20),\n",
    "    },\n",
    "    'selpreterm': {\n",
    "        'cta': np.linspace(\n",
    "            0.05, 1, num=20),\n",
    "        'ctc': np.linspace(\n",
    "            0.5, 10, num=20),\n",
    "        'ctt': np.linspace(\n",
    "            0.25, 5, num=20),\n",
    "    }\n",
    "}\n",
    "\n",
    "# find the location of all leucine codons to convert leu codon serial number\n",
    "# to absolute position along yfp in codon units for simulation\n",
    "leupositions = dict()\n",
    "leucodon_number = 1\n",
    "for position in range(0, len(yfp0), 3):\n",
    "    currentcodon = yfp0[position:position + 3]\n",
    "    if currentcodon == 'CTG':\n",
    "        leupositions[leucodon_number] = position / 3\n",
    "        leucodon_number += 1\n",
    "\n",
    "for model in stallstrengthranges:\n",
    "    for index in range(20):\n",
    "        fitdata = list()\n",
    "        for mutant in mutation_locations:\n",
    "            if len(mutant) > 1:\n",
    "                raise \"Only single mutants allowed\"\n",
    "            codon = mutant.keys()[0]\n",
    "            pos = mutant.values()[0]\n",
    "            fitdata.append({\n",
    "                'codon': codonDict[codon.upper()],\n",
    "                'pos': leupositions[pos],\n",
    "                'stallstrength': stallstrengthranges[model][codon][index]\n",
    "            })\n",
    "        fitdata = pd.DataFrame(fitdata)\n",
    "        fitdata.to_csv(\n",
    "            '../annotations/simulations/run2/{0}_stallstrengthindex_{1}.tsv'.\n",
    "            format(model, index),\n",
    "            sep='\\t',\n",
    "            index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Simulation run 2: Reporter simulation with systematically varying duration of ribosome stalling during leucine starvation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting simulation_run_2.py\n"
     ]
    }
   ],
   "source": [
    "%%writefile simulation_run_2.py\n",
    "#!/usr/bin/env python\n",
    "# SBATCH --mem=8000\n",
    "\n",
    "import subprocess as sp\n",
    "import os\n",
    "import sys\n",
    "import itertools\n",
    "import numpy as np\n",
    "\n",
    "jobindex = int(sys.argv[1])\n",
    "currentindex = -1\n",
    "\n",
    "allfiles = os.listdir('../annotations/simulations/run2/')\n",
    "\n",
    "mrnafiles = list(filter(lambda x: x.startswith('yfp'), allfiles))\n",
    "mrnafiles = ['../annotations/simulations/run2/' + File for File in mrnafiles]\n",
    "\n",
    "terminationRates = {\n",
    "    'trafficjam': {'--5prime-preterm-rate': 0},\n",
    "    '5primepreterm': {'--5prime-preterm-rate': 1},\n",
    "    'selpreterm': {'--selective-preterm-rate': 1},\n",
    "}\n",
    "\n",
    "for (mrnafile,\n",
    "     typeOfTermination\n",
    "     ) in list(itertools.product(mrnafiles, terminationRates)):\n",
    "    currentindex += 1\n",
    "    if currentindex != jobindex:\n",
    "        continue\n",
    "    termoption = terminationRates[typeOfTermination].keys()[0]\n",
    "    termvalue = terminationRates[typeOfTermination].values()[0]\n",
    "    stallstrengthfiles = list(\n",
    "        filter(lambda x: x.startswith(typeOfTermination), allfiles))\n",
    "    stallstrengthfiles = [\n",
    "        '../annotations/simulations/run2/' + File for File in stallstrengthfiles]\n",
    "    for stallstrengthfile in stallstrengthfiles:\n",
    "        cmd = ' '.join([\n",
    "            './reporter_simulation',\n",
    "            '--trna-concn', '../annotations/simulations/leucine.starvation.average.trna.concentrations.tsv',\n",
    "            termoption, '%0.2f' % termvalue,\n",
    "            '--threshold-accommodation-rate', '22',\n",
    "            '--output-prefix', '../rawdata/simulations/run2/',\n",
    "            '--stall-strength-file', stallstrengthfile,\n",
    "            '--input-genes', mrnafile\n",
    "        ])\n",
    "        sp.check_output(cmd, shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import subprocess as sp\n",
    "# loop submits each simulation to a different node of the cluster\n",
    "for index in range(75):\n",
    "    sp.check_output([\n",
    "        'sbatch',  # for SLURM cluster; this line can be commented out if running locally\n",
    "        '-t',\n",
    "        '300',  # for SLURM cluster; this line can be commented out if running locally\n",
    "        '-n',\n",
    "        '1',  # for SLURM cluster; this line can be commented out if running locally\n",
    "        'simulation_run_2.py',\n",
    "        str(index)\n",
    "    ])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create mRNA sequences for simulation run 13 with systematically varying tRNA accommodation rate (stall duration) at Ser codons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "mutation_locations = [\n",
    "    {\n",
    "        'tcg': 2\n",
    "    },\n",
    "    {\n",
    "        'tcg': 3\n",
    "    },\n",
    "    {\n",
    "        'tcg': 4\n",
    "    },\n",
    "    {\n",
    "        'tcg': 5\n",
    "    },\n",
    "    {\n",
    "        'tcg': 6\n",
    "    },\n",
    "    {\n",
    "        'tcg': 7\n",
    "    },\n",
    "]\n",
    "\n",
    "yfpmutants = dict()\n",
    "for mutant in mutation_locations:\n",
    "    key = '_'.join(['yfp'] + [\n",
    "        codon + str(location) for codon, location in mutant.items()\n",
    "    ])\n",
    "    yfpmutants[key] = list(yfp_agc)\n",
    "    codon_number = 0\n",
    "    for position in range(0, len(yfp_agc), 3):\n",
    "        currentcodon = yfp_agc[position:position + 3]\n",
    "        # proceed only if the codon is a Leu codon (which are all CTG in yfp0)\n",
    "        if currentcodon not in ['AGC']:\n",
    "            continue\n",
    "        codon_number += 1\n",
    "        for codon in mutant.keys():\n",
    "            if codon_number == mutant[codon]:\n",
    "                yfpmutants[key][position:position + 3] = codon.upper()\n",
    "    yfpmutants[key] = ''.join(yfpmutants[key])\n",
    "\n",
    "defaultMrnaCopyNumber = 1  # per cell\n",
    "defaultInitationRate = 0.3  # s-1, This is the median initiation rate\n",
    "\n",
    "listOfInitiationRates = [defaultInitationRate]\n",
    "\n",
    "for initiationRate in listOfInitiationRates:\n",
    "    for mutant in yfpmutants:\n",
    "        outputFile = '../annotations/simulations/run13/' + \\\n",
    "            '%s_initiationrate_%0.4f.csv' % (mutant, initiationRate)\n",
    "        num_seq = ''.join(\n",
    "            get_numerical_codon_sequence(yfpmutants[mutant][:-3]))\n",
    "        File = open(outputFile, 'w')\n",
    "\n",
    "        File.write(\"%0.4f\\t%d\\t%s\\n\" %\n",
    "                   (initiationRate, defaultMrnaCopyNumber,\n",
    "                    get_numerical_codon_sequence(yfp_agc[:-3])))\n",
    "\n",
    "        File.write(\"%0.4f\\t%d\\t%s\\n\" % (initiationRate, defaultMrnaCopyNumber,\n",
    "                                        num_seq))\n",
    "    File.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create varying tRNA accommodation rate (stall duration) at Ser codons for simulation run 13"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "stallstrengthranges = {\n",
    "    'trafficjam': {\n",
    "        'tcg': np.linspace(\n",
    "            0.02, 0.4, num=20),\n",
    "    },\n",
    "    '5primepreterm': {\n",
    "        'tcg': np.linspace(\n",
    "            0.05, 1, num=20),\n",
    "    },\n",
    "    'selpreterm': {\n",
    "        'tcg': np.linspace(\n",
    "            0.5, 10, num=20),\n",
    "    }\n",
    "}\n",
    "\n",
    "# find the location of all leucine codons to convert leu codon serial number\n",
    "# to absolute position along yfp in codon units for simulation\n",
    "serpositions = dict()\n",
    "sercodon_number = 1\n",
    "for position in range(0, len(yfp_agc), 3):\n",
    "    currentcodon = yfp_agc[position:position + 3]\n",
    "    if currentcodon == 'AGC':\n",
    "        serpositions[sercodon_number] = position / 3\n",
    "        sercodon_number += 1\n",
    "\n",
    "for model in stallstrengthranges:\n",
    "    for index in range(20):\n",
    "        fitdata = list()\n",
    "        for mutant in mutation_locations:\n",
    "            if len(mutant) > 1:\n",
    "                raise \"Only single mutants allowed\"\n",
    "            codon = mutant.keys()[0]\n",
    "            pos = mutant.values()[0]\n",
    "            fitdata.append({\n",
    "                'codon': codonDict[codon.upper()],\n",
    "                'pos': serpositions[pos],\n",
    "                'stallstrength': stallstrengthranges[model][codon][index]\n",
    "            })\n",
    "        fitdata = pd.DataFrame(fitdata)\n",
    "        fitdata.to_csv(\n",
    "            '../annotations/simulations/run13/{0}_stallstrengthindex_{1}.tsv'.\n",
    "            format(model, index),\n",
    "            sep='\\t',\n",
    "            index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Simulation run 13: Reporter simulation with systematically varying duration of ribosome stalling during serine starvation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing simulation_run_13.py\n"
     ]
    }
   ],
   "source": [
    "%%writefile simulation_run_13.py\n",
    "#!/usr/bin/env python\n",
    "# SBATCH --mem=8000\n",
    "\n",
    "import subprocess as sp\n",
    "import os\n",
    "import sys\n",
    "import itertools\n",
    "import numpy as np\n",
    "\n",
    "jobindex = int(sys.argv[1])\n",
    "currentindex = -1\n",
    "\n",
    "allfiles = os.listdir('../annotations/simulations/run13/')\n",
    "\n",
    "mrnafiles = list(filter(lambda x: x.startswith('yfp'), allfiles))\n",
    "mrnafiles = ['../annotations/simulations/run13/' + File for File in mrnafiles]\n",
    "\n",
    "terminationRates = {\n",
    "    'trafficjam': {'--5prime-preterm-rate': 0},\n",
    "    '5primepreterm': {'--5prime-preterm-rate': 1},\n",
    "    'selpreterm': {'--selective-preterm-rate': 1},\n",
    "}\n",
    "\n",
    "for (mrnafile,\n",
    "     typeOfTermination\n",
    "     ) in list(itertools.product(mrnafiles, terminationRates)):\n",
    "    currentindex += 1\n",
    "    if currentindex != jobindex:\n",
    "        continue\n",
    "    termoption = terminationRates[typeOfTermination].keys()[0]\n",
    "    termvalue = terminationRates[typeOfTermination].values()[0]\n",
    "    stallstrengthfiles = list(\n",
    "        filter(lambda x: x.startswith(typeOfTermination), allfiles))\n",
    "    stallstrengthfiles = [\n",
    "        '../annotations/simulations/run13/' + File for File in stallstrengthfiles]\n",
    "    for stallstrengthfile in stallstrengthfiles:\n",
    "        cmd = ' '.join([\n",
    "            './reporter_simulation',\n",
    "            '--trna-concn', '../annotations/simulations/serine.starvation.average.trna.concentrations.tsv',\n",
    "            termoption, '%0.2f' % termvalue,\n",
    "            '--threshold-accommodation-rate', '22',\n",
    "            '--output-prefix', '../rawdata/simulations/run13/',\n",
    "            '--stall-strength-file', stallstrengthfile,\n",
    "            '--input-genes', mrnafile\n",
    "        ])\n",
    "        sp.check_output(cmd, shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import subprocess as sp\n",
    "# loop submits each simulation to a different node of the cluster\n",
    "for index in range(21):\n",
    "    sp.check_output([\n",
    "        'sbatch',  # for SLURM cluster; this line can be commented out if running locally\n",
    "        '-t',\n",
    "        '300',  # for SLURM cluster; this line can be commented out if running locally\n",
    "        '-n',\n",
    "        '1',  # for SLURM cluster; this line can be commented out if running locally\n",
    "        'simulation_run_13.py',\n",
    "        str(index)\n",
    "    ])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
